### PR TITLE
Ensure there are no races in ThreadDeathWatcher, related to [#6062]

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We used a MPSC queue in ThreadDeathWatcher and checked if it empty via isEmpty() from multiple threads if very unlucky. Depending on the implementation this is not safe and may even produce things like live-locks.

Modification:

- Change to use a MPSP queue
- Some simplifications

Result:

No more risk to run into issues when multiple threads call watch(...) / unwatch(...) concurrently.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch FYI... I was thinking how to improve the `awaitInactivity` to be less racy but could not come up with anything good.

File: common/src/main/java/io/netty/util/ThreadDeathWatcher.java [hunk 1]
```diff
@@ -142,22 +146,19 @@ public void run() {
...
                 }
 
                 if (watchees.isEmpty() && pendingEntries.isEmpty()) {
+                    Thread currentThread = Thread.currentThread();
 
                     // Mark the current worker thread as stopped.
                     // The following CAS must always success and must be uncontended,
                     // because only one watcher thread should be running at the same time.
-                    boolean stopped = started.compareAndSet(true, false);
+                    boolean stopped = watcherThread.compareAndSet(currentThread, null);
```
Comment (netkins) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: MAJOR Remove this useless assignment to local variable "stopped". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

File: common/src/main/java/io/netty/util/ThreadDeathWatcher.java [hunk 2]
```diff
@@ -121,10 +125,10 @@ public static boolean awaitInactivity(long timeout, TimeUnit unit) throws Interr
             throw new NullPointerException("unit");
         }
 
-        Thread watcherThread = ThreadDeathWatcher.watcherThread;
-        if (watcherThread != null) {
-            watcherThread.join(unit.toMillis(timeout));
-            return !watcherThread.isAlive();
+        Thread t = watcherThread.get();
+        if (t != null) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: The synchronization approach used by this class is racy:

1. Watcher#run detects there are no more pending entries, so sets watcher thread to `null`.
2. schedule puts something on the `pendingEntries` queue.
3. `awaitInactivity` is called.
4. schedule or Watcher#run sets the watcher thread.

`awaitInactivity` may exit prematurely.

The comments provide a loop hole that this method should only be called after all calls to `watch`. However it makes me a bit uneasy that the code responsible for watching all threads is itself racy. I wonder if we can have a `private static void Thread TERMINATED = ...;` or some token object that is respected to avoid scheduling new work after the thread death watcher is "terminated" (e.g. `awaitInactivity` is called).

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: @Scottmitch I don't think so... This would be a breaking change. Maybe we should just deprecate this method and add a new one that does this. WDYT ?

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: IMHO the existing behavior is a bug. The javadocs state the expected behavior of the interface, but the code fails to enforce the behavior. My vote is just to fix the code and remove the race condition.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: @Scottmitch where does the javadoc state that you are now allowed to add more stuff after this method returns ? Disallowing this is a breakage IMHO.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: The javadocs on `awaitInactivity`:

```
this operation is only useful when you want to ensure that the watcher thread is terminated
<strong>after</strong> your application is shut down and there's no chance of calling
{@link #watch(Thread, Runnable)} afterwards.
```

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: Which says nothing about that it will have side effects on the future calls to add / remove methods.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: Thats why I think we should deprecate it and add another method which adds these side-effects.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: Legitime in the sense that it is not forbidden in terms of the API documentation. I don't think it makes a lot of sense tho.. I just would hate to break some app because of such a change and I think we could easily just add a new method that will exactly do this and add this behaviour.

Something like `awaitInactivityAndClose()`

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: Are you suggesting leaving the existing code as is? I think that is what I'm struggling with. The existing implementation of `awaitInactivity` is racy and therefore unreliable. I don't yet see the value in preserving this unreliable behavior.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: Not sure I am suggestion to keep it like it is but at least not alter it to have side-effects.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: I am comfortable if we want to offer the following functionality:

- `awaitInactivity` - will block until all currently scheduled jobs are complete, with no race, also allows schedule to be called after this method is called.
- `awaitInactivityAndClose` - will block until all currently scheduled jobs are complete, with no race, all subsequent calls to watch will fail.

The "no race" part being the key :)

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: @Scottmitch ok sounds good

File: common/src/main/java/io/netty/util/ThreadDeathWatcher.java [hunk 3]
```diff
@@ -116,18 +121,28 @@ private static void schedule(Thread thread, Runnable task, boolean isWatch) {
...
-            throw new NullPointerException("unit");
-        }
-
-        Thread watcherThread = ThreadDeathWatcher.watcherThread;
-        if (watcherThread != null) {
-            watcherThread.join(unit.toMillis(timeout));
-            return !watcherThread.isAlive();
-        } else {
-            return true;
+    public static boolean awaitInactivity(long amount, TimeUnit unit) throws InterruptedException {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: doesn't this have the same race condition described in https://github.com/netty/netty/pull/6104#discussion_r91187728?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: @Scottmitch need to think about this a bit more... That said let me factor out the Queue replacement to another PR as I think we want to pull this in in all cases.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ThreadDeathWatcher.java]: @Scottmitch see https://github.com/netty/netty/pull/6149

**Comment by normanmaurer** [comment]
let me close this for now and come-up with something better once I have time
